database-commons/src/main/java/co/cask/SchemaReader.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask;
+
+import co.cask.cdap.api.data.schema.Schema;
+
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Main Interface to read db specific types.
+ */
+public interface SchemaReader {
+
+  List<Schema.Field> getSchemaFields(@Nullable String schemaStr) throws SQLException;
+
+  List<Schema.Field> getSchemaFields() throws SQLException;
+
+  Schema getSchema(ResultSetMetaData metadata, int index) throws SQLException;
```

Comment:
your readers already have resultSet in implementation. Why do you need to additionally pass metadata here?